File: 0079-word-search/0079-word-search.py
from typing import List

class Solution:
    def exist(self, board: List[List[str]], word: str) -> bool:
        rows, cols = len(board), len(board[0])
        word_len = len(word)
        visited = [[False for _ in range(cols)] for _ in range(rows)]

        def helper(index, i, j):
            # Base cases
            if index == word_len:  # All characters matched
                return True
            if i < 0 or j < 0 or i >= rows or j >= cols:  # Out of bounds
                return False
            if visited[i][j] or board[i][j] != word[index]:  # Already visited or mismatch
                return False

            # Mark as visited
            visited[i][j] = True

            # Explore in all four directions
            found = (
                helper(index + 1, i, j - 1) or  # Left
                helper(index + 1, i, j + 1) or  # Right
                helper(index + 1, i - 1, j) or  # Up
                helper(index + 1, i + 1, j)     # Down
            )

            # Backtrack
            visited[i][j] = False
            return found

        # Start the search from every cell
        for i in range(rows):
            for j in range(cols):
                if board[i][j] == word[0]:  # Potential starting point
                    if helper(0, i, j):
                        return True
        return False

File: 0079-word-search/0079-word-search.java
class Solution {
    public static boolean check(int index, int i, int j, String word, char[][] board, int [][] visited){
        int size = word.length();
        int row = board.length, col = board[0].length;
        if(index == size){
            return true;
        }
        if(i < 0 || i >= row || j < 0 || j >= col){
            return false;
        }
        if(visited[i][j] == 1 || word.charAt(index) != board[i][j]){
            return false;
        }
        visited[i][j] = 1;
        boolean left = check(index + 1, i , j - 1, word, board, visited);
        boolean right = check(index + 1, i , j + 1, word, board, visited);
        boolean up = check(index + 1, i - 1 , j, word, board, visited);
        boolean down = check(index + 1, i + 1 , j, word, board, visited);
        visited[i][j] = 0;
        return left || right || up || down;

    }
    public boolean exist(char[][] board, String word) {
        int row = board.length;
        int col = board[0].length;
        int index = 0;
        char ch  = word.charAt(0);
        int [][] visited = new int [row][col];

        for(int i = 0 ; i < row; i++){
            for(int j = 0; j < col; j++){
                visited[i][j] = 0;
            
            }
            }

        for(int i = 0 ; i < row; i++){
            for(int j = 0; j < col; j++){
                if(board[i][j] == ch){
                    if(check(0,i,j,word,board, visited)){
                        return true;
                    }
                }
            }
        }
        return false;
    }
}